### PR TITLE
Feature NXP-29529 Elastic Upgrade to 7.9.2

### PR DESCRIPTION
**Pull request by bdelbosc**
Upgrade to Elastic 7.9.1.
Not to be merged immediately because CI pipeline needs to be updated.
Still, the code can be reviewed.

### CODE REVIEW

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestESClient.java [hunk 1]
```diff
@@ -97,12 +97,13 @@ public void closeESClient() throws Exception {
     @Test
     public void testIndexOptimisticConcurrency() {
         client.createIndex("foo-index", "{\"index\" : {\"number_of_shards\" : 1, \"number_of_replicas\" : 0}}");
+        // client.createIndex("foo-index", "{\"settings\":{\"index\" : {\"number_of_shards\" : 1, \"number_of_replicas\" : 0}}}");
```
Comment (kevinleturc) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestESClient.java:100]: To remove?

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/hint/GeoShapeESHintQueryBuilder.java [hunk 2]
```diff
@@ -40,10 +40,10 @@
...
     public QueryBuilder make(EsHint hint, String fieldName, Object value) {
-        String[] values = validate(value, 4, "shapeId, type, index and path");
-        return QueryBuilders.geoShapeQuery(fieldName, values[0], values[1])
+        String[] values = validate(value, 3, "shapeId, index and path");
+        return QueryBuilders.geoShapeQuery(fieldName, values[0])
                             .relation(ShapeRelation.WITHIN)
-                            .indexedShapeIndex(values[2])
-                            .indexedShapePath(values[3]);
+                            .indexedShapeIndex(values[1])
+                            .indexedShapePath(values[2]);
```
Comment (kevinleturc) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/hint/GeoShapeESHintQueryBuilder.java:47]: Shouldn't we keep a backward compatibility mechanism for the existing queries with 4 parameters written by users?

**Review by kevinleturc — COMMENTED** [review]
Great work 👍. I'm wondering how the CI has succeeded without changing the Elasticsearch version, is there some kind of compatibility mode?

**Comment by bdelbosc** [comment]
> I'm wondering how the CI has succeeded without changing the Elasticsearch version, is there some kind of compatibility mode?
No, this version cannot work with elastic 6.x, my guess is that the PR CI pipeline only uses embedded Elasticsearch.

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/client/ESRestClient.java [hunk 3]
```diff
@@ -321,7 +323,8 @@ public DeleteResponse delete(DeleteRequest request) {
     public SearchResponse search(SearchRequest request) {
         try (Scope ignored = getScopedSpan("elastic/_search", request.toString())) {
             return client.search(request, RequestOptions.DEFAULT);
-        } catch (IOException e) {
+        } catch (IOException|ElasticsearchStatusException e) {
```
Comment (ataillefer) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/client/ESRestClient.java:326]:
```suggestion
        } catch (IOException | ElasticsearchStatusException e) {
```

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchServiceImpl.java [hunk 4]
```diff
@@ -20,13 +20,12 @@
...
 package org.nuxeo.elasticsearch.core;
 
-import static org.nuxeo.elasticsearch.ElasticSearchConstants.DOC_TYPE;
-
 import java.util.List;
 
 import org.apache.commons.lang3.StringUtils;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
+import org.apache.lucene.search.TotalHits;
```
Comment (ataillefer) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchServiceImpl.java:28]: really needed?

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestMapping.java [hunk 5]
```diff
@@ -23,13 +23,15 @@
 
 import javax.inject.Inject;
 
+import org.elasticsearch.ElasticsearchStatusException;
```
Comment (ataillefer) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestMapping.java:26]: not needed I guess

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/main/java/org/nuxeo/elasticsearch/audit/ESAuditBackend.java [hunk 6]
```diff
@@ -715,9 +715,11 @@ protected void ensureUIDSequencer(ESClient esClient) {
...
                                                 .aggregation(AggregationBuilders.max("maxAgg").field("id")));
         SearchResponse searchResponse = esClient.search(request);
-        Max agg = searchResponse.getAggregations().get("maxAgg");
-        long maxLogEntryId = (long) agg.getValue();
-
+        Aggregation agg = searchResponse.getAggregations().get("maxAgg");
+        long maxLogEntryId = 0;
+        if (agg.getMetadata() != null && agg.getMetadata().containsKey(Aggregation.CommonFields.VALUE)) {
+            maxLogEntryId = (long) agg.getMetadata().get(Aggregation.CommonFields.VALUE);
+        }
```
Comment (ataillefer) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/main/java/org/nuxeo/elasticsearch/audit/ESAuditBackend.java:722]:
```suggestion
        long maxLogEntryId = 0;
        Aggregation agg = searchResponse.getAggregations().get("maxAgg");
        Map<String, Object> aggMetadata = agg.getMetaData();
        if (aggMetadata != null) {
            maxLogEntryId = (long) aggMetadata.getOrDefault(Aggregation.CommonFields.VALUE, 0);
        }
```

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchIndexingImpl.java [hunk 7]
```diff
@@ -290,7 +290,7 @@ void processIndexCommand(IndexingCommand cmd) {
             return;
         }
         if (log.isDebugEnabled()) {
-            logDebugMessageTruncated(String.format("Index request: curl -XPUT 'http://localhost:9200/%s/%s/%s' -d '%s'",
+            logDebugMessageTruncated(String.format("Index request: curl -XPUT 'http://localhost:9200/%s/%s' -d '%s'",
```
Comment (ataillefer) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchIndexingImpl.java:293]: to be clear, this should be in the same commit as https://github.com/nuxeo/nuxeo/pull/4293/commits/f2afb3d1cc87dd9b40c7e2d50499c7af381dd55e#diff-f0d6271e8fca3291a638cf9797c75ddaL295

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchIndexingImpl.java [hunk 8]
```diff
@@ -360,7 +360,7 @@ void processDeleteCommandRecursive(IndexingCommand cmd) {
             // Build bulk delete request
             BulkRequest bulkRequest = new BulkRequest();
             for (SearchHit hit : response.getHits().getHits()) {
-                bulkRequest.add(new DeleteRequest(hit.getIndex(), hit.getType(), hit.getId()));
+                bulkRequest.add(new DeleteRequest(hit.getIndex(), hit.getId()));
```
Comment (ataillefer) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchIndexingImpl.java:363]: and this too probably

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/resources/log4j2-test.xml [hunk 9]
```diff
@@ -12,6 +12,10 @@
   </Appenders>
   <Loggers>
     <Logger name="org.nuxeo" level="info" />
+    <!-- Remove WARN on elastic embedded because dangling indices cannot be detected -->
+    <Logger name="org.elasticsearch.gateway.DanglingIndicesState" level="error" />
+    <!-- Remove WARN on elastic embedded because node and cluster identifiers change between test suite -->
```
Comment (ataillefer) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/resources/log4j2-test.xml:17]:
```suggestion
    <!-- Remove WARN on elastic embedded because node and cluster identifiers change between test suites -->
```

File: pom.xml [hunk 10]
```diff
@@ -69,8 +69,8 @@
     <jaxb-api.version>2.3.3</jaxb-api.version>
     <jaxb.version>2.3.3</jaxb.version>
     <tomcat.version>9.0.38</tomcat.version>
-    <lucene.version>8.6.0</lucene.version>
-    <elasticsearch.version>7.9.0</elasticsearch.version> <!-- when changing this make sure JNA version is in sync -->
+    <lucene.version>8.6.2</lucene.version>
+    <elasticsearch.version>7.9.1</elasticsearch.version> <!-- when changing this make sure JNA version is in sync -->
```
Comment (ataillefer) [inline pom.xml:73]: in the end, should be squashed with https://github.com/nuxeo/nuxeo/pull/4293/commits/f2afb3d1cc87dd9b40c7e2d50499c7af381dd55e#diff-600376dffeb79835ede4a0b285078036L72

**Review by ataillefer — COMMENTED** [review]
Nice job :muscle: 

Missed only one duplicate contrib: https://github.com/nuxeo/nuxeo/blob/master/modules/platform/nuxeo-drive-server/nuxeo-drive-elasticsearch/src/test/java/org/nuxeo/drive/test/ESAuditFeature.java#L45

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/hint/GeoShapeESHintQueryBuilder.java (hunk 2 above)
Comment (bdelbosc) [reply modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/hint/GeoShapeESHintQueryBuilder.java:47]: yes, you are right, I will add the "type" back here saying it is unused.

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/main/java/org/nuxeo/elasticsearch/audit/ESAuditBackend.java (hunk 6 above)
Comment (bdelbosc) [reply modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/main/java/org/nuxeo/elasticsearch/audit/ESAuditBackend.java:722]: I find it harder to read

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchIndexingImpl.java (hunk 8 above)
Comment (bdelbosc) [reply modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchIndexingImpl.java:363]: this too what?

File: pom.xml (hunk 10 above)
Comment (bdelbosc) [reply pom.xml]: yes, I have updated now upgrade is on 7.9.2

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchIndexingImpl.java (hunk 7 above)
Comment (bdelbosc) [reply modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/core/ElasticSearchIndexingImpl.java:293]: will do that better next time

**Review by ataillefer — DISMISSED** [review]

**Review by ataillefer — APPROVED** [review]

**Comment by bdelbosc** [comment]
code has just been re-align to master

**Review by kevinleturc — APPROVED** [review]

**Comment by bdelbosc** [comment]
:wave: PR has been aligned again to master (no change) and build.groovy is ready for merge (no more version of nuxeo-helm-chart and runtime unit test enabled again see commit 90fa8cf025415f65c12b0d7a9dd09dec0ef77779 the only update since your last review).
Note that the build is not going to pass until nuxeo-helm-chart is merged and available.

**Review by kevinleturc — APPROVED** [review]

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/main/java/org/nuxeo/elasticsearch/audit/ESAuditBackend.java [hunk 11]
```diff
@@ -715,8 +715,11 @@ protected void ensureUIDSequencer(ESClient esClient) {
         request.source(new SearchSourceBuilder().query(QueryBuilders.matchAllQuery())
                                                 .aggregation(AggregationBuilders.max("maxAgg").field("id")));
         SearchResponse searchResponse = esClient.search(request);
-        Max agg = searchResponse.getAggregations().get("maxAgg");
-        long maxLogEntryId = (long) agg.getValue();
+        Aggregation agg = searchResponse.getAggregations().get("maxAgg");
+        long maxLogEntryId = 0;
+        if (agg.getMetadata() != null && agg.getMetadata().containsKey(Aggregation.CommonFields.VALUE)) {
+            maxLogEntryId = (long) agg.getMetadata().get(Aggregation.CommonFields.VALUE);
+        }
```
Comment (efge) [inline modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/main/java/org/nuxeo/elasticsearch/audit/ESAuditBackend.java:722]: Could we put `agg.getMetaData()` in a local `var` at least to avoid repetition? It could be done directly with

    var aggMetadata = searchResponse.getAggregations().get("maxAgg").getMetaData();

But I would also really want to advocate `getOrDefault`, it's a habit that's good to have.

File: pom.xml [hunk 12]
```diff
@@ -69,8 +69,8 @@
     <jaxb-api.version>2.3.3</jaxb-api.version>
     <jaxb.version>2.3.3</jaxb.version>
     <tomcat.version>9.0.39</tomcat.version>
-    <lucene.version>7.5.0</lucene.version>
-    <elasticsearch.version>6.5.3</elasticsearch.version> <!-- when changing this make sure JNA version is in sync -->
+    <lucene.version>8.6.2</lucene.version>
+    <elasticsearch.version>7.9.2</elasticsearch.version> <!-- when changing this make sure JNA version is in sync -->
```
Comment (efge) [inline pom.xml:73]: Note that 7.9.3 is out now.

Comment (bdelbosc) [reply pom.xml:73]: yes, it is hard to follow their release cycle, I prefer to continue on 7.9.2 for now,
I think there is going to be a 7.10 as well

File: modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/main/java/org/nuxeo/elasticsearch/audit/ESAuditBackend.java (hunk 11 above)
Comment (bdelbosc) [reply modules/platform/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/main/java/org/nuxeo/elasticsearch/audit/ESAuditBackend.java:722]: ok will do that another time, the CI cannot validate a new change before merging

**Comment by bdelbosc** [comment]
code aligned to master again to resolve conflict with "NXP-29729: Upgrade org.json:json dependency"

**Review by kevinleturc — APPROVED** [review]

**Review by troger — APPROVED** [review]
